### DISCUSSION TITLE
Creating a counter with decimals

### DISCUSSION THREAD

**Discussion by Claponi** [olikraus/u8g2#2499] [Q&A] [answered]
Hi,
using MUI_U8G2 on arduino I'm trying to create a counter with two variables, one of them would be decimal of the other. 
That way when varB = 9+1, varA =varA+1.
I'm using a rotary encoder (same code that MUIInputSimpleRotary example).
Of course it's not working as I expect...

This is the part of my code about it :

> uint8_t DMX_Ref_Value_A = 0;
> uint8_t DMX_Ref_Value_B = 0; //decimal part
> 
> uint8_t MyIncremental_mui_u8g2_u8_min_max_wm_mud_pi(mui_t *ui, uint8_t msg)
> {
>   uint8_t result = mui_u8g2_u8_min_max_wm_mud_pi(ui, msg);  // call original function
>   if (msg == MUIF_MSG_VALUE_INCREMENT)
>   {
>       if (DMX_Ref_Value_B == 9 && DMX_Ref_Value_A < 100){
>       DMX_Ref_Value_A = DMX_Ref_Value_A + 1;
>       }
>   }
>   return result;
> 
>  MUIF_U8G2_U8_MIN_MAX("AF", &DMX_Ref_Value_A, 0, 100, mui_u8g2_u8_min_max_wm_mud_pi),
>  MUIF_U8G2_U8_MIN_MAX("AG", &DMX_Ref_Value_B, 0, 9, MyIncremental_mui_u8g2_u8_min_max_wm_mud_pi), 
> 

Any help would be much appreciated.

**Comment by olikraus** [comment]
Is there any reason why uint8 input can't be used?
https://github.com/olikraus/u8g2/wiki/muiref#numbers-uint8

Just set the range from 0 to 99

**Reply by Claponi** [reply]
Thank you for your reply.

The idea is then to do something like :
varC = varA + varB/10 // where  0.0%<=varC<=100.0%

That way I have a more precise percentage value to control my DMX output (here light power).
If I do only from 0% to 100% it is not precise enough when mapping into 8bits DMX, and it is not usual for the user to input 0-255 value when working with DMX in cinema industry.  Also, some time you use 16bits DMX to code light value.

**Reply by olikraus** [reply]
ah, understand.
Usually in mud mode you need to listen to the event messages:
https://github.com/olikraus/u8g2/blob/761f8c2eb3ec589cfc39207f9038b85be8c451f5/sys/arduino/u8g2_page_buffer/MUIWaveformSimpleRotary/MUIWaveformSimpleRotary.ino#L547-L571
You also need to test for "9" before you call the original function.

**Reply by olikraus** [reply, marked as answer]
Unfortunately it is a little bit more complicated, because you need to consider some edge cases like 100.0 --> 0.0 rollover.

I created a full example here: https://github.com/olikraus/u8g2/blob/master/sys/arduino/u8g2_page_buffer/MUIFractionSimpleRotary/MUIFractionSimpleRotary.ino

The increment case for example is here:
https://github.com/olikraus/u8g2/blob/9cc8dbc582d89316aa9caa83b78e01428150517f/sys/arduino/u8g2_page_buffer/MUIFractionSimpleRotary/MUIFractionSimpleRotary.ino#L414-L429

**Reply by Claponi** [reply]
thank you very much !
I'll try your MUIFractionSimpleRotary.ino
Cause now I'm stuck with "event_next" being triggered when going from decimal MUI to another MUI...

**Reply by Claponi** [reply]
Well this is perfect !! thank you very much  for your help.

**Reply by olikraus** [reply]
There is one more issue: If the decimal part is moved to 9, then you can still change the pre-decimal value to 100. Then you get 100.9%.
Maybe we can "misuse" the draw event to force the decimal value to 0.

**Reply by olikraus** [reply]
This will fix the 100.x% problem:
https://github.com/olikraus/u8g2/blob/5a1779d0d095ea5e3eac196fe64c53114b29dba9/sys/arduino/u8g2_page_buffer/MUIFractionSimpleRotary/MUIFractionSimpleRotary.ino#L414-L418

**Reply by olikraus** [reply]
Comments fixed: 
https://github.com/olikraus/u8g2/blob/3ed61ebfd36317ca4d101447f35577e764011f38/sys/arduino/u8g2_page_buffer/MUIFractionSimpleRotary/MUIFractionSimpleRotary.ino#L414-L418

**Comment by Claponi** [comment]
As I'm not using the rollover (strange for a light to bright from max to 0), this is what I did. 
It also solved the 100.x% issue :

> uint8_t muif_decimal(mui_t *ui, uint8_t msg)
> {
>   uint8_t return_value = 0; 
>   switch(msg)
>   {
>     case MUIF_MSG_EVENT_NEXT:                           // for mud mode
>       if ( ui->is_mud == 0 ) {
>         return_value = mui_u8g2_u8_min_max_wm_mud_pi(ui, msg);                            // none mud state
>       }
>       else if ( DMX_RefValue_Decimal == 9 && DMX_RefValue_PreDecimal < 100 ) {                                                                                        // 46.9 --> 47.0
>         DMX_RefValue_PreDecimal++;
>         return_value = mui_u8g2_u8_min_max_wm_mud_pi(ui, msg);                            // let MUI modify the DMX_RefValue_Decimal value
>       }
>       else if ( DMX_RefValue_Decimal == 0 && DMX_RefValue_PreDecimal == 100 ) {                                                // 100.0 stuck at 100.0
>         DMX_RefValue_PreDecimal = 100;
>         return_value = 1;               // don't call the original function
>       }
>       else {                                                                                                                      // all other cases
>         return_value = mui_u8g2_u8_min_max_wm_mud_pi(ui, msg);                            // let MUI modify the DMX_RefValue_Decimal value
>       }
>       break;
>     case MUIF_MSG_EVENT_PREV:                           // for mud mode
>       if ( ui->is_mud == 0 ) {
>         return_value = mui_u8g2_u8_min_max_wm_mud_pi(ui, msg);                            // none mud state
>       }
>       else if ( DMX_RefValue_Decimal == 0 ) {                                                                                        // x.0
>         if ( DMX_RefValue_PreDecimal == 0 )                                                                                         // 0.0 stuck at 0.0
>         {
>           DMX_RefValue_PreDecimal = 0;
>           return_value = 1;               // don't call the original function
>         }
>         else {                                                                                                                      // 47.0 --> 46.9
>           DMX_RefValue_PreDecimal--;
>           return_value = mui_u8g2_u8_min_max_wm_mud_pi(ui, msg);                            // let MUI modify the DMX_RefValue_Decimal value
>         }
>       }
>       else {
>         return_value = mui_u8g2_u8_min_max_wm_mud_pi(ui, msg);                            // let MUI modify the DMX_RefValue_Decimal value
>       }
>       break;
>     default:
>       return_value = mui_u8g2_u8_min_max_wm_mud_pi(ui, msg);                    // for any other messages, just call the original MUIF
>   }
>   return return_value;
> }

**Comment by Claponi** [comment]
And glad to your help I think I've a better understanding of your library, and coding in general (I'm a true beginner :).

so thank you again !
